Individual Claim

Dinonard opened the pull request:
**Pull Request Summary**

Main purpose of this commit is to solve scalability problems with the former `claim` implementation.
Instead of using group claims, each staker will now be responsible for claiming their own rewards.

This entails changes in storage schema and other extrinsic calls.
* dApp state is now tracked using a dedicated enum `DAppState`
* Cleaned `Forcing` enum of unused values
* `EraStakingPoints` now contains info about staked value, number of stakers and whether dapp rewards were claimed for an era
* `StakerInfo` is a new struct which handles `(staker, contract)` info storage in a scalable manner
  * the idea is to wrap history into a vector of limited size which prevents stakers from overflowing our DB
  * methods are available for easier handling, extensive and thorough UTs provided
 * Introduced `RewardHandling` enum as preparation for future task (to avoid storage migration)
 * `HistoryDepth` limit has been removed
 * `BonusEraDuration` has been removed as it serves no purpose anymore
 * New extrinsic `withdraw_from_unregistered` to cleanup stakes after contract was removed
 * claiming functionality now has separate calls for `claim_staker` and `claim_dapp`
 * very extensive UTs provided

**TODO**
- [x] rename EraStakeAndReward to something more appropriate
- [x] manual test on local node
- [x] update precompiles code
- [x] update documentation (README)
- [x] re-run all benchmarks prior to merge
- [x] implement new benchmarking
- [x] implement & test storage migration
- [x] cover code with UTs
- [x] implement individual claim 

**Check list**
- [x] contains breaking changes
- [x] adds new feature
- [x] modifies existing feature (bug fix or improvements)
- [ ] relies on other tasks
- [ ] documentation changes
- [x] tests and/or benchmarks are included
- [ ] changed API client type definition or chain metadata

Maar-io commented on frame/dapps-staking/src/lib.rs:
```diff
@@ -97,6 +130,128 @@ impl Default for Version {
+            self.stakes.push(EraStake::new(value, current_era))
+        } else {
+            let era_stake = self.stakes.last().unwrap(); // exists if vec not empty
+            if era_stake.era > current_era {
```
When can this happen?

Dinonard replied:
Initially, I implemented `stake` and `unstake` without this check.
But since we have to be 100% sure nothing can get messed up, I added this check.

To answer your question - it cannot happen :).
It's just a safety measure in case I have a bug in the code that would allow it.

Maar-io commented on frame/dapps-staking/src/lib.rs:
```diff
@@ -97,6 +130,128 @@ impl Default for Version {
+
+        let new_stake_value = era_stake.staked.saturating_sub(value);
+
+        if current_era == era_stake.era {
+            *self.stakes.last_mut().unwrap() = EraStake::new(new_stake_value, current_era)
+        } else {
+            self.stakes
+                .push(EraStake::new(new_stake_value, current_era))
+        }
```
code duplication? same is in stake()

Dinonard replied:
Good catch!
I tried to avoid this by wrapping it up in a method but the problem is I cannot make it private.
And I don't want to expose this as part of the API.

E.g. it would be something like 'update_staked_value(era, value)`

hoonsubin commented on frame/dapps-staking/src/benchmarking.rs:
```diff
@@ -180,22 +195,37 @@ benchmarks! {
+        let (developer, contract_id) = register_contract::<T>()?;
 
-        let number_of_stakers = n - 1;
+        let number_of_stakers = 5;
```
Can you tell me why the number of stakers are set as a constant?

Dinonard replied:
Leftover from previous implementation when weight was dependent on the number of active stakers.
I could change it to e.g. 1 or just remove this (immutable) variable.

Maar-io commented on frame/dapps-staking/src/lib.rs:
```diff
@@ -97,6 +130,128 @@ impl Default for Version {
+        } else {
+            let era_stake = self.stakes[0]; // not empty so it exists
+
+            if self.stakes.len() == 1 || self.stakes[1].era > era_stake.era + 1 {
```
What if `self.stakes[1]` does not exist?

Dinonard replied:
That code will never execute if the first part of the clause is true.

If `self.is_empty()` is false and `self.stakes.len() == 1` is false, it means that `self.stakes.len()` is greater than 1.
So `self.stakes[1]' will always exist if that part of the code is executed.

akru replied:
Could you explain this condition, in that case `self.stakes[1].era > era_stake.era + 1` is not true?

Dinonard replied:
We have two possible branches:
* first one is when we only have a single element in the vec OR if the next era for which we have info isn't the consecutive era which we are claiming. E.g.:
<1, 1000>, <5, 2000> ==> <2, 1000>, <5, 2000>
*In this scenario, we just modify the **first** element by changing its era
* the second scenario can only be the case where the next element in vec is the consecutive era
* this means we can just delete the first element since that information isn't relevant anymore
<1, 1000>, <2, 2000> ==> <2, 2000>

Maar-io commented on frame/dapps-staking/src/lib.rs:
```diff
@@ -196,12 +351,39 @@ where
+
+impl Default for RewardHandling {
+    fn default() -> Self {
+        RewardHandling::PayoutAndStake
```
Shouldn't the default be `OnlyPayout`?

Dinonard replied:
We can discuss this.

My motivation is that we should make `PayoutAndStake` default because this is what benefits users more, as well as the ecosystem.

This can be changed in the commit that will add actual support for this.

Maar-io commented on frame/dapps-staking/src/lib.rs:
```diff
@@ -24,6 +27,8 @@ mod mock;
+mod tests_lib;
 
 pub use pallet::pallet::*;
 pub use sp_staking::SessionIndex;
```
Unused. you can delete

Maar-io commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -371,14 +387,26 @@ pub mod pallet {
         /// Warning: After this action contract can not be assigned again.
-        #[pallet::weight(T::WeightInfo::unregister(T::MaxNumberOfStakersPerContract::get()))]
+        #[pallet::weight(T::WeightInfo::unregister())]
         pub fn unregister(
```
once we get rid of Sudo, how can an admin do unregister for contracts that prove to not be worked on or malicious?

Dinonard replied:
Sure, I will change it to root-only.
We will also need a couple for root-only force methods in addition to this.

Maar-io commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -371,14 +387,26 @@ pub mod pallet {
+                Error::<T>::NotOwnedContract,
+            );
+
+            // TODO: Is this really needed? Why shouldn't same account be eligible to register multiple contracts?
```
perhaps spam protection?

Dinonard replied:
We have two levels of spam protection:
- dev needs to be whitelisted
- each registration requires reservation amount

It'd be a very expensive spam.

Maar-io commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -389,39 +417,68 @@ pub mod pallet {
+        /// Withdraw locked funds from a contract that was unregistered.
+        /// Funds don't need to undergo the unbonding period - they are returned immediately.
+        #[pallet::weight(T::WeightInfo::withdraw_from_unregistered())]
+        pub fn withdraw_from_unregistered(
```
@bobo-k2 how do you plan to implement this on frontend?

akru replied:
@Dinonard could you explain please why it should be dedicated method? It looks like a case for general unbond method.

Dinonard replied:
It's actually quite different - the checks differ.

Here we require that all pending rewards have been claimed. Also, there's no unbonding period here.

Sure, I could put it all into `unbond_and_unstake` but it creates a more complicated method with many **if-s**.
This, in my opinion, is much better. Code is easier to test and there is less room for error.

akru replied:
Clean interface is also important. Not sure why withdraw from unregistered dapps should be different from general withdraw.

Dinonard replied:
I agree with you in that statement but I don't want to make code more complicated - by doing this split, there is no branching, most if/else clauses are avoided.

And the behavior is inherently different.

If we use `unbond_and_unstake` we can `withdaraw_unbonded` after unbonding period has passed.
However, `withdraw_from_unregistered` acts as `unbond_unstake_and_withdraw` all at once.
So the behavior is quite different.

There is no unbonding period here - funds are returned directly to the user.
Also, you cannot call this if there are pending rewards - this is one of the simplifications I made to avoid complexity in the `StakerInfo`.

Should we maybe rename it?

Maar-io commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -625,105 +681,136 @@ pub mod pallet {
-        /// Any user can call this function.
-        #[pallet::weight(T::WeightInfo::claim(T::MaxNumberOfStakersPerContract::get() + 1))]
-        pub fn claim(
+        // TODO: do we need to add force methods or at least methods that allow others to claim for someone else?
```
once we implement re-stake, projects might be interested to claim in name of the stakers and thus increase the staked amount for their contracts

Dinonard replied:
Sure, I was thinking the same thing.

It's easy to add a wrapped call, no problem there.

Dinonard replied:
I can implement wrappers after I get review from Aleks.

Maar-io commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -625,105 +681,136 @@ pub mod pallet {
+            // Ensure we have something to claim
+            let mut staker_info = Self::staker_info(&staker, &contract_id);
+            let (era, staked) = staker_info.claim();
+            ensure!(staked > Zero::zero(), Error::<T>::NotStakedContract);
```
`staked != Zero::zero()` maybe?

Dinonard replied:
I'd prefer `!staked.is_zero()` but the when we got audited by Quantstamp, they explicitly told me to prefer **is greater than** check.

Maar-io commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -625,105 +681,136 @@ pub mod pallet {
+            let reward_and_stake =
+                Self::era_reward_and_stake(era).ok_or(Error::<T>::UnknownEraReward)?;
+
+            let (_, stakers_reward) = Self::dev_stakers_split(
```
rename to `stakers_joint_reward `

Maar-io commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -625,105 +681,136 @@ pub mod pallet {
+
+        /// Claim earned dapp rewards for the specified era.
+        #[pallet::weight(T::WeightInfo::claim_dapp())]
+        pub fn claim_dapp(
```
I'd still stick for a single claim call (stakers + dev) for the front-end but it is up to you and @bobo-k2

Dinonard replied:
For the frontend, I agree. Not part of this commit though.

Dinonard commented on frame/dapps-staking/src/lib.rs:
```diff
@@ -193,39 +229,77 @@ impl<Balance: AtLeast32BitUnsigned + Copy> StakerInfo<Balance> {
-        let new_stake_value = era_stake.staked.saturating_sub(value);
+            let new_stake_value = era_stake.staked.saturating_sub(value);
+            if current_era == era_stake.era {
+                *self.stakes.last_mut().unwrap() = EraStake::new(new_stake_value, current_era)
```
Improve this too

akru commented on frame/dapps-staking/src/lib.rs:
```diff
@@ -101,11 +130,239 @@ impl<AccountId: Ord, Balance: HasCompact + Default> Default
+    /// Repeated calls would continue to modify vector following the same rule as in *4.*
+    ///
+    fn claim(&mut self) -> (EraIndex, Balance) {
+        if self.is_empty() {
```
Let's use `if let` here too.

Dinonard replied:
I did before but didn't like how it looked...but I'll push the change again.

akru commented on frame/dapps-staking/src/lib.rs:
```diff
@@ -101,11 +130,239 @@ impl<AccountId: Ord, Balance: HasCompact + Default> Default
+
+    /// Removes unstaked values if they're no longer valid for comprehension
+    fn clean_unstaked(&mut self) {
+        if !self.stakes.is_empty() && self.stakes[0].staked.is_zero() {
```
May be we should include this logic into method that use it, it's not too heavy to use dedicated method but readability will increase if we move this operations into methods instead of alias.

Dinonard replied:
Sure, no problem.

akru commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -178,7 +196,19 @@ pub mod pallet {
+
+    #[pallet::storage]
+    #[pallet::getter(fn staker_info)]
+    pub(crate) type StakersInfo<T: Config> = StorageDoubleMap<
```
May be rename `StakersInfo` -> `GeneralStakerInfo`, because of possible types `Stakers` = `Staker`

Dinonard replied:
Sure, I can do that.

@impelcrypto please not that I will change this in the new version.

akru commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -372,57 +416,87 @@ pub mod pallet {
-
-            let registered_contract =
-                RegisteredDevelopers::<T>::get(&developer).ok_or(Error::<T>::NotOwnedContract)?;
+            Self::ensure_pallet_enabled()?;
```
I'm not sure that root calls should have this guard

Dinonard replied:
This is intentional - I will use maintenance mode to prevent any storage access during storage migration.
`unregister` is also just a normal extrinsic.

If there's a need to use `unregister` during maintenance mode (imo there shouldn't be), we can just wrap it into a batch call where we `disable maintenance - execute unregister - enable maintenance`

Dinonard replied:
Btw, perhaps looking at the test utility method for `withdraw_from_unregistered` would help. It think it shows how behavior is different.

codespool commented on frame/dapps-staking/README.md:
```diff
@@ -3,13 +3,13 @@ This document describes the interface for the pallet-dapps-staking.
-1. [Storage](#Storage)
-1. [Referent implementatio](#Referent)
-1. [FAQ](#FAQ)
+2. [Types](#Types)
```
Why keep the table of contents if there's no content anymore?

Dinonard replied:
Good catch, I will remove it.

Dinonard committed:
Individual claim

codespool commented:
Found this while fixing my own PR:
you're still using reward_handling property name in migrations:
https://github.com/AstarNetwork/astar-frame/blob/6df33856f0616fcecf21e5bb593a444961c9bbe5/frame/dapps-staking/src/migrations.rs#L548
is that intentional?

Dinonard commented:
> Found this while fixing my own PR: you're still using reward_handling property name in migrations:
> 
> https://github.com/AstarNetwork/astar-frame/blob/6df33856f0616fcecf21e5bb593a444961c9bbe5/frame/dapps-staking/src/migrations.rs#L548
> 
> is that intentional?

Oh, I didn't rename the variable name. I will do it now.
The type is `RewardDestination` though, even here.

codespool commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -775,11 +873,31 @@ pub mod pallet {
+        #[pallet::weight(T::WeightInfo::maintenance_mode())]
+        pub fn maintenance_mode(
+            origin: OriginFor<T>,
+            enable_maintenace: bool,
```
typo :)

Dinonard replied:
fixed

Dinonard committed:
Small changes

Dinonard committed:
No-loop dapps-staking (#23)

* Removal of iter from dapps-staking

Address comments

Benchmark and precompiles update

Improve UT

* correct weight calculation

* Fix precompile compilation issue

Maar-io committed:
Use pallet precompile-utils for pallet-precompile-dapps-staking (#22)

* added register()

bond_and_stake()

bond and unbond

claim and test

remove util.rs

cleanup and test

fmt fix

added log

* read_staked_amount_ss58

* fmt fix

* accountId decode - not working

* use AccountId32 in mock

* rebase to 0.9.17

* fmt after rebase to 0.9.17

* use Bytes instead of U256

* test fix for ss58

* single method for ss58 and h160 read_staked_amount

* added register()

bond_and_stake()

bond and unbond

claim and test

remove util.rs

cleanup and test

fmt fix

added log

read_staked_amount_ss58

fmt fix

accountId decode - not working

use AccountId32 in mock

rebase to 0.9.17

fmt after rebase to 0.9.17

use Bytes instead of U256

test fix for ss58

single method for ss58 and h160 read_staked_amount

* updates for individual claim

* renaming staking_info()

* fmt fix

* rtenaming contract_stake_info

Maar-io reviewed (approved):
LGTM

Dinonard commented:
@codespool You can rebase on the default branch now. No more force-pushes :D
